### PR TITLE
Add geolocate option to 'hosts' command

### PR DESCRIPTION
**Pull request by cbrnrd**
Addressing #8192 (finally). This PR adds the ability to perform an IP-based geolocation using the [geoip gem](https://github.com/cjheath/geoip). The gem uses a local database (stored in `data/GeoLiteCity.dat`) to fetch the IP information. Using the `-g` flag for the `hosts` command, the user is able to determine lot's of information about a given IP address. Thanks to @busterb for this idea!

Note: the ip address doesn't necessarily need to be stored in `hosts`, but it can be
## Verification

List the steps needed to make sure this thing works

- [x] Make sure travis gives the go ahead
- [x] Fix any problems using multiple hosts (working on that now)

#### Single target

```
msf > hosts -g 8.8.8.8
[*] Info for 8.8.8.8
    IP: 8.8.8.8
    Continent: NA
    Country: United States
    Region: California
    City: Mountain View
    Longitude: -122.0838 (about)
    Latitude: 37.385999999999996 (about)

msf > 
```

#### Multiple hosts (and an invalid one)

```
msf > hosts -g 8.8.8.8 999.999.999.999 8.8.4.4
[*] Info for 8.8.8.8
    IP: 8.8.8.8
    Continent: NA
    Country: United States
    Region: California
    City: Mountain View
    Longitude: -122.0838 (about)
    Latitude: 37.385999999999996 (about)

[-] 999.999.999.999 is not a valid IP address

[*] Info for 8.8.4.4
    IP: 8.8.4.4
    Continent: NA
    Country: United States
    Region: 
    City: 
    Longitude: -97.822 (about)
    Latitude: 37.751000000000005 (about)
```

### CODE REVIEW

**Comment by wvu-r7** [comment]
What's the license on `GeoLiteCity.dat`?

**Comment by thecarterb** [comment]
[Creative Commons Attribution-ShareAlike 4.0 International License](http://dev.maxmind.com/geoip/legacy/geolite/#License)

**Comment by wvu-r7** [comment]
That was a rhetorical question. Please satisfy the attribution requirement. :)

**Comment by thecarterb** [comment]
I've added the MaxMind attribution and the geoip license. Should I add the (very long) Creative Commons Attribution-ShareAlike 4.0 International License to `LICENSE`?

File: lib/msf/ui/console/command_dispatcher/db.rb [hunk 1]
```diff
@@ -382,6 +382,32 @@ def cmd_hosts(*args)
...
+        # RegEx#match? seems to be faster on newer versions of Ruby
+        if !Rex::Socket::MATCH_IPV4.match?(host) && !Rex::Socket::MATCH_IPV6.match?(host)
+          if Rex::Socket::MATCH_IPV4_PRIVATE.match?(host)
+            print_error("You cannot geolocate local IP addresses")
+            return
+          end
+          print_error("#{host} is not a valid IP address")
+          return
+        end
+        glcdat = "#{File.expand_path(File.dirname(__FILE__))}/../../../../../data/GeoLiteCity.dat"  # Point to data directory
```
Comment (void-in) [inline lib/msf/ui/console/command_dispatcher/db.rb]: The data directory can be referenced like:
```ruby
glcdat = ::File.join(Msf::Config::data_directory, 'GeoCity.dat')
```

**Comment by thecarterb** [comment]
I've added support for multiple hosts. Anything else that I should change?

File: lib/msf/ui/console/command_dispatcher/db.rb (hunk 1 above)
Comment (void-in) [inline lib/msf/ui/console/command_dispatcher/db.rb]: @thecarterb Sorry for the mistake. The file name should be ```GeoLiteCity.dat``` and not ```GeoCity.dat```. Please correct the file name in the updated commit.

**Comment by thecarterb** [comment]
Already done :)

File: lib/msf/ui/console/command_dispatcher/db.rb (hunk 1 above)
Comment (thecarterb) [inline lib/msf/ui/console/command_dispatcher/db.rb]: Already done :)

**Comment by wvu-r7** [comment]
@thecarterb: Yeah, `LICENSE` is what I had in mind, but IANAL. :)

**Comment by thecarterb** [comment]
I suppose if I have the attribution, then the requirement is satisfied (i hope).

**Comment by egypt** [comment]
The ShareAlike clause sounds an awful lot like GPL. I'm super reluctant to add it on those grounds.

**Comment by thecarterb** [comment]
@egypt Probably because it's fairly similar to it (copyleft lisense). Straight from the website: alt-text

**Comment by dmaloney-r7** [comment]
I don't believe this is a feature that appeals to a necessarily large portion of the user base, and adding a large-ish local database in service of it seems somewhat counterproductive. There is also some question about licensing concerns involved. I would suggest the following solution to @thecarterb :

Remake this as a Metasploit plugin, and offer it up on your own. That way people who really want this features can go and download your plugin and use it, without impinging on anyone who does not want it.

**Comment by thecarterb** [comment]
@dmaloney-r7 Will do :)

**Pull request closed without merging**